truffle/src/main/java/org/jruby/truffle/nodes/rubinius/IOBufferPrimitiveNodes.java

```diff
@@ -0,0 +1,193 @@
+
+        public IOBufferUnshiftPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public IOBufferUnshiftPrimitiveNode(IOBufferUnshiftPrimitiveNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public int unshift(VirtualFrame frame, RubyBasicObject ioBuffer, RubyString string, int startPosition) {
+            rubyWithSelf(frame, ioBuffer, "@write_synced = false");
+
+            int stringSize = string.getByteList().realSize() - startPosition;
+            final int usedSpace = (int) rubyWithSelf(frame, ioBuffer, "@used");
+            final int availableSpace = IOBUFFER_SIZE - usedSpace;
+
+            if(stringSize > availableSpace) {
+                stringSize = availableSpace;
+            }
+
+            ByteList storage = ((RubiniusByteArray) rubyWithSelf(frame, ioBuffer, "@storage")).getBytes();
+
+            // Data is copied here - can we do something COW?
+            System.arraycopy(string.getByteList().unsafeBytes(), startPosition, storage.getUnsafeBytes(), usedSpace, stringSize);
+
+            rubyWithSelf(frame, ioBuffer, "@used = used", "used", usedSpace + stringSize);
+
+            return stringSize;
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "iobuffer_fill")
+    public static abstract class IOBufferFillPrimitiveNode extends RubiniusPrimitiveNode {
+
+        public IOBufferFillPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public IOBufferFillPrimitiveNode(IOBufferFillPrimitiveNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public int fill(VirtualFrame frame, RubyBasicObject ioBuffer, RubyBasicObject io) {
+            int fd = (int) rubyWithSelf(frame, io, "@descriptor");
+
+            // TODO CS 21-Apr-15 allocating this buffer for each read is crazy
+            byte[] readBuffer = new byte[STACK_BUF_SZ];
+            int count = STACK_BUF_SZ;
+
+            if(left(frame, ioBuffer) < count) {
+                count = left(frame, ioBuffer);
+            }
+
+            int bytesRead;
+
+            while (true) {
+                bytesRead = getContext().getPosix().read(fd, readBuffer, count);
+
+                if (bytesRead == -1) {
+                    final int errno = getContext().getPosix().errno();
+
+                    if (errno == Errno.ECONNRESET.intValue() || errno == Errno.ETIMEDOUT.intValue()) {
+                        // Treat as seeing eof
+                        bytesRead = 0;
+                        break;
+                    } else if (errno == Errno.EAGAIN.intValue() || errno == Errno.EINTR.intValue()) {
+                        //if (!state -> check_async(calling_environment))
+                        //    return NULL;
+                        //io -> ensure_open(state);
+                        getContext().getSafepointManager().poll(this);
```

Comment:
No, I'm afraid it doesn't appear to work at the moment. Can you interrupt all reads in a process somehow?